src/poetry/core/version/markers.py

```diff
@@ -611,6 +646,55 @@ def intersect(self, other: BaseMarker) -> BaseMarker:
     def union(self, other: BaseMarker) -> BaseMarker:
         return union(self, other)
 
+    def intersect_simplify(self, other: BaseMarker) -> BaseMarker | None:
+        """
+        Finds a couple of easy simplifications for intersection on MarkerUnions:
+
+            - intersection with any marker that appears as part of the union is just
+              that marker
+
+            - intersection between two markerunions where one is contained by the other
+              is just the smaller of the two
+
+            - intersection between two markerunions where there are some common markers
+              and the intersection of unique markers is not a single marker
+        """
+        if other in self._markers:
+            return other
+
+        if isinstance(other, MarkerUnion):
+            our_markers = set(self.markers)
+            their_markers = set(other.markers)
+
+            if our_markers.issubset(their_markers):
+                return self
+
+            if their_markers.issubset(our_markers):
+                return other
+
+            if not our_markers.intersection(their_markers):
+                return None
+
+            # use lists instead of sets for deterministic order
+            common_markers = [
+                marker for marker in self.markers if marker in other.markers
+            ]
+
+            if common_markers:
+                unique_markers = [
+                    marker for marker in self.markers if marker not in common_markers
+                ]
+                other_unique_markers = [
+                    marker for marker in other.markers if marker not in common_markers
+                ]
+                unique_intersection = MarkerUnion(*unique_markers).intersect(
+                    MarkerUnion(*other_unique_markers)
+                )
+                if isinstance(unique_intersection, SingleMarker):
```

Comment:
> You could restore symmetry by testing `if isinstance(unique_union, (SingleMarker, AnyMarker)):` in the other place

In the other place we can do better by testing `if not isinstance(unique_union, MarkerUnion):`. Doing the same with `MultiMarker` in this place is not possible because we always prefer dnf:

https://github.com/python-poetry/poetry-core/blob/5a273c3c7c84e8140773e242d9d4f51647f8585d/src/poetry/core/version/markers.py#L775-L784

Reply:
> In the other place we can do better by testing if not isinstance(unique_union, MarkerUnion)

I don't think this is doing better, though, because it's never possible to achieve the other types.  ie I think that the two different formulations are equivalent, one of them just restores symmetry.

Reply:
By the way I am not absolutely 100% mathematically certain of the claim that `unique_union` can't be a `MultiMarker` - but it is at any rate difficult enough to make it happen that I can't figure out how to do it (and also rare enough that the unit tests don't achieve it either).  

Comparing the value of the possible micro-optimisation in case a `MultiMarker` does occur here against the value of slightly prettier code - I'd find it hard to make a strong case either way.  Personally I am drawn to the symmetry but no big deal.

Reply:
`unique_union` should be a `MultiMarker` for something like
```
unique_markers = {'python_version >= "3.7"', 'python_version < "3.11"'}
other_unique_markers = {'python_version >= "3.8"', 'python_version < "3.12"'}
```
However, due to the nature of `dnf`, I believe there will always be another marker with only one of the unique markers so the markers will be merged later anyway. As you said it's probably just a micro-optimization.

I just realized that allowing `MultiMarker` would require to use lists instead of sets for `unique_markers` and `other_unique_markers` to keep deterministic results. Maybe, that's what tips the scales. So let's choose symmetry for now until there is a good reason to choose the opposite.